File: src/Java/algorithms/mathandprob/AddBinary.java
package algorithms.mathandprob; 

import java.util.*; 

public class AddBinary {
    /* Function: Add two numbers that present in binary. 
     * Solution: Can be a template. 
     */ 
    
    public String addBinary(String a, String b) {
        if (a == null || a.length() == 0) {
            return b; 
        }
        if (b == null || b.length() == 0) {
            return a; 
        }
        
        StringBuffer sb = new StringBuffer(); 
        int pt1 = a.length() - 1, pt2 = b.length() - 1; 
        int c = 0; 
        while (pt1 >= 0 || pt2 >= 0 || c > 0) {
            int val1 = pt1 >= 0 ? a.charAt(pt1) - '0' : 0;
            int val2 = pt2 >= 0 ? b.charAt(pt2) - '0' : 0; 
            int num = c + val1 + val2; 
            sb.insert(0, num % 2); 
            c = num / 2; 
            --pt1; 
            --pt2; 
        }
        
        return sb.toString(); 
    }
}

File: src/CPP/src/leetcode/AddBinary.cpp
/*
 * [Source] https://leetcode.com/problems/add-binary/
 * [Difficulty]: Easy
 * [Tag]: Math
 * [Tag]: String
 */

#include <iostream>

using namespace std;

// [Solution]: Use a carry
// [Corner Case]: two numbers are not same length
class Solution {
};

/* Java solution
https://github.com/ZhenyingZhu/ClassicAlgorithms/blob/master/src/algorithms/mathandprob/AddBinary.java
 */

/* Java solution
public class Solution {
    public String addBinary(String a, String b) {
		if(a.length()<b.length()){ // a always longer
			String tmp=a;
			a=b;
			b=tmp;
		}
		char[] first=a.toCharArray();
		char[] second=b.toCharArray();
		int aLen=a.length();
		int bLen=b.length();
		int[] res=new int[aLen];
		int c=0;
		for(int i=0; i<bLen; i++){
			res[aLen-1-i]=first[aLen-1-i]+second[bLen-1-i]-2*'0'+c;
			if(res[aLen-1-i]>=2){
				res[aLen-1-i]-=2;
				c=1;
			}else{
				c=0;
			}
		}
		for(int i=aLen-bLen-1; i>=0; i--){
			res[i]=first[i]-'0'+c;
			if(res[i]>=2){
				res[i]-=2;
				c=1;
			}else{
				c=0;
			}
		}
		StringBuffer result=new StringBuffer();
		if(c==1) result.append('1');
		for(int i=0; i<res.length; i++){
			result.append(Integer.toString(res[i]));
		}
		return result.toString();
    }
}
 */

int main() {
    Solution sol;

    return 0;
}
